### ISSUE TITLE
dataclass attributes not inherited, but only when parent dataclass is in a different distribution?

### ISSUE THREAD

**Issue by davidatbu** [microsoft/pyright#974]
**Describe the bug**
Say dataclass `A` is in some package `foo_pkg`. Another dataclass `B` inherits from `A`. Pyright doesn't allow one to initialize `B` with arguments for the attributes that were defined in `A`.

**To Reproduce**
A minimal repro that requires just a clone and running a script can be had [here](https://github.com/davidatbu/pyright_bug_repro). Here's the important part:

```python
from foo_pkg import A
from dataclasses import dataclass


@dataclass
class B(A):
    attr_in_b: int


B(attr_in_a=0, attr_in_b=1)
```

**Expected behavior**
This is a false positive by Pyright.

**VS Code extension or command-line**
The repro linked uses the command-line version. My pyright version is 1.1.65.

**Additional context**
It's worth nothing, this doesn't happen if the class `A` is in the same file, or in a file that is available because python adds the working dir to `sys.path`. Only when `A` is in an "installed" package(for lack of a better word), does this happen.

**Comment by erictraut** [comment]
Thanks for the clear repro steps.

The problem here is that the import of `A` is failing. If you run "bar.py" in the python interpreter, you will see that it throws an exception:
```
Traceback (most recent call last):
  File "bar.py", line 1, in <module>
    from foo_pkg import A
ImportError: cannot import name 'A' from 'foo_pkg' (unknown location)
```

Pyright likewise indicates that the import isn't resolved:
```
"A" is unknown import symbol [1, 21]
No parameter named "attr_in_a" [10, 3]
```

If you change your code to the following, it will work:
```
from foo_pkg.foo_pkg import A
```

Alternatively, you can delete the nested `foo_pkg` directory and move its contents up to the top-level directory of the same name.

**Comment by davidatbu** [comment]
Thanks for the quick reply! 

As I said in the original post, this *only happens when `foo_pkg` is an installed package*. (The reason that I had a nested `foo_pkg/` directory with a `setup.py` on the outside is because that is a [very typical structure for a package source](https://github.com/psf/requests) ).

When you tried to reproduce the bug, are you sure you ran `pip install foo_pkg/`? If so, python shouldn't have had a problem finding the package `foo_pkg`. Also, pyright doesn't give the error you showed when I do `pip install`, it  complains only  about the attribute.

To make what I mean clear: Here's a slightly annotated bash session that I just ran.

```bash
$ git clone git@github.com:davidatbu/pyright_bug_repro.git && cd pyright_bug_repro/

$ python -m venv /tmp/venv &&  source /tmp/venv/bin/activate # create and activate a virtualenv

$ pip install wheel # Install wheel as usual

$ pip install foo_pkg/ # Don't forget the slash at the end
 
$ pip freeze
foo-pkg==0.1
$ rm foo_pkg/ -r
$ tree . # Show dir contents
.
├── bar.py
├── README.md
└── repro.sh

0 directories, 3 files
(venv) davidat@davidat-laptop:~/git/pyright_bug_repro$ python bar.py # no complaints from Python
(venv) davidat@davidat-laptop:~/git/pyright_bug_repro$ pyright bar.py
stubPath /home/davidat/git/pyright_bug_repro/typings is not a valid directory.
Assuming Python platform Linux
Searching for source files
Found 1 source file
/home/davidat/git/pyright_bug_repro/bar.py
  10:3 - error: No parameter named "attr_in_a" (reportGeneralTypeIssues)
1 error, 0 warnings
Completed in 0.563sec
```

**Comment by erictraut** [comment]
If I follow your instructions, I am able to repro the error if "useLibraryCodeForTypes" is disabled. Since you don't have a stub file for the installed package, Pyright doesn't know the type of `A`. If you enable "useLibraryCodeForTypes", it's able to use the type information from the installed package.

The logic that synthesizes dataclass constructors should probably allow any parameters if one or more of the base classes are of an unknown type. That will completely eliminate all type checking for these constructors, but it will eliminate false positive errors in this case.

This change will be in the next release of Pyright.

**Comment by davidatbu** [comment]
Thanks @erictraut  for looking into this some more!

`foo_pkg` in the repro I gave is a PEP 561 compliant package because it includes a `py.typed ` marker file. PEP 561 specifies that it aims to support "package maintainers who would like to add type information inline" (ie, not in stub files). Here's a quote from the PEP

>  if there is nothing overriding the installed package, and it opts into type checking, inline types SHOULD be used.

Indeed, it seems that we've discovered a bigger issue in pyright, which is that it doesn't comply the above part of PEP 561. To confirm this, I included the line `reveal_type(A.attr_in_a)` in `bar.py` in the repro and pyright gave the error

```
warning: Type of "A.attr_in_a" is "Unknown"
````

While that type should be `int`.

**Comment by erictraut** [comment]
Yes, we're having an ongoing debate internally about how to handle PEP 561. The problem is that PEP 561 provides too many options and doesn't provide enough guidance for package developers. It also doesn't specify what "typed" means and how complete types need to be before it's fine to claim that a package is "typed". As a spec, it leaves a lot to be desired, and it probably needs to evolve.

In practice, packages that include a 'py.typed' often contain incomplete types. The difference between a package that claims to be typed and one that doesn't claim to be typed turns out to be not that meaningful in many cases. The 'py.typed' marker is an inaccurate heuristic at best, which is why we currently ignore it.

By contrast, packages that ship with stubs (or ship companion stub packages) tend to be much more complete, and there are good tools available to verify the completeness of stubs. 

We are discussing the idea of changing the behavior to honor the 'py.typed' marker. One thought is to change `useLibraryCodeForTypes` into a tri-state setting — one that allows for "never", "always" and "only if py.typed marker is present".

**Comment by davidatbu** [comment]
> An ongoing debate.

I see. FWIW, when I put on my package developer hat, I really value the option to explicitly opt-in to type-checking(without having to resort to writing stub files or creating a stub-only package). So I'd definitely vote for the third "only if py.typed marker is present" option.

Also, I believe the current README claims that pyright implements PEP 561, so I think a small but noticable section detailing such an (IMO quite important) exception would be quite helpful.

**Comment by erictraut** [comment]
Thanks, I appreciate your feedback and perspective.

**Comment by davidatbu** [comment]
Thank you for pyright!

**Comment by erictraut** [comment]
The fix is in Pyright 1.1.66, which I just published.

As for the point you made about py.typed and conformance with PEP 561, we made some progress on this topic, and we should have a final plan shortly. I think you'll be happy with the conclusion.

**Comment by davidatbu** [comment]
Thanks for the fix Eric!

Also excited to hear what y'all got in the works regarding PEP 561 conformance.
